### ISSUE TITLE
Request body JSON matcher (or custom matchers)

### ISSUE THREAD

**Issue by drbr** [httptoolkit/mockttp#27]
I want to create a mock that matches a request only if it has certain things in its JSON body (e.g. `_.isMatch(request, partialContents)`). The recently-added regex body matcher lets me match a single simple field in the body, but I want to match more complicated structures.

I can see three potential ways to be able to do this.
1. Have mockttp define such a matcher
2. Define my own custom matcher on the MockRuleBuilder
3. I could do something fancy in `thenCallback`, but this seems less clean

Are any of these possible?

**Comment by pimterry** [comment]
Yes, I think those are your current options. I agree that it's not perfect, I would like to have something better implemented, and I'd definitely accept PRs for this if you're interested.

The two main routes I can see for Mockttp changes are:

1. Create a callback-based matching function (discussed in #19), something like `get('/').matching((request) => ...).thenReply(...)`
2. Create a specific matcher for this case, something like
  `get('/').withJSONMatching({ a: 1 }).thenReply(...)`

Option 1 has the benefit of being more generic, and we're definitely going to need it anyway. It's non-trivial though (for browsers/remote clients, you need communication between the server & browser for each request that could match, in a sort-of similar way to `thenCallback()`), and comes with a performance footgun. I.e. your matching function is going to be called against every single request going through the mock, so if you accidentally add an expensive matching and you're mocking/proxying a fair bit of traffic it'll bite, and that gets worse if we allow async like thenCallback does, and gets worse still for remote clients (which need a websocket RTT for every potential match).

Option 2 avoids those problems, it's very quick and easy to implement & easy to use imo, but doing that in general eventually ends up with an explosion of matchers. Not a big problem, but a little messy.

Any strong preferences? In the short-term, to me it sounds like option 2 (implementing `withJSONMatching(o)`) probably makes the most sense for you. It'd be good to also add `withJSON` for exact matches while we're there. What do you think?

**Comment by drbr** [comment]
Hi, thanks for the quick reply!

Yes, the predicate matcher seems very powerful and it would definitely be nice to have. Since it accepts an arbitrary function, it seems like every matcher could be implemented using it, so it essentially gives users the capability to define custom matchers (which appears to be currently not possible). It appears simple to me, but I don't understand the HTTP subtleties (regarding "live communication") that you say are required, so I'm not going to try to build it.

Adding JSON matchers does seem very straightforward and safe, and I would gladly use them if they were to exist. If you were able to magically conjure these matchers in the short term, that would make me very happy, and I might be able to put together a PR in the less-short term.

I've also been thinking about bypassing the lack of custom matcher entirely by writing my own matching logic in `thenCallback`, which would also allow me to dynamically construct the result based on the request body. I haven't tried it out yet, but I'm optimistic about the power that it would unlock.

**Comment by hanvyj** [comment]
Yeah, we had this requirement too. The work around using `thenCallback` is very functional though. Here's what we're doing:

```
  const endpoint = mockServer.put(path).thenCallback(actualRequest => {
    if (
      expectedBody === undefined ||
      (actualRequest.body.json &&
        isMatch(expectedBody, actualRequest.body.json))
    ) {
      return {
        status: response.status,
        json: response.body
      };
    } else {
      // the request body doesn't match the expected
      return {
        status: 503
      };
    }
  });
```

**Comment by drbr** [comment]
@hanvyj Yep, that's essentially the code I ended up writing. It's working like a charm :)

So, no _need_ for the json or predicate matcher, but if it were available, I might eventually rewrite my code to use it.

**Comment by pimterry** [comment]
@hanvyj @drbr, it's your lucky day: I had a little time spare this afternoon, so I've implemented `withJsonBody` (exact match) and `withJsonBodyIncluding` (fuzzy match).

See https://github.com/httptoolkit/mockttp/commit/d15316d9fae5a3765c017b02f299537935665a50 for full details. Now released as version 0.14.6, enjoy :smiley:.
